### DISCUSSION TITLE
CORS middleware does not reject invalid requests

### DISCUSSION THREAD

**Discussion by Fogapod** [fastapi/fastapi#7360] [Questions]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

```Python
from fastapi import FastAPI
from fastapi.middleware.cors import CORSMiddleware

app = FastAPI()
app.add_middleware(
    CORSMiddleware,
    allow_origins=["http://localhost"],
    allow_credentials=True,
    allow_methods=["GET"],
    allow_headers=["*"],
)

# POST endpoint is registered, but it should not be accessible
# because allow_methods=["GET"]
@app.post("/")
async def hello():
    return {"message": "Hello World"}
```

```sh
~ uvicorn main:app --port 8081
```

### Description

I have multiple issues with this minimal example:
- Requests with wrong/missing `Origin` header are able to go through:
```sh
~ curl -X POST localhost:8081 -H 'Origin: http://remotehost' -v
*   Trying 127.0.0.1:8081...
* Connected to localhost (127.0.0.1) port 8081 (#0)
> POST / HTTP/1.1
> Host: localhost:8081
> User-Agent: curl/7.71.1
> Accept: */*
> Origin: http://remotehost
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< date: Mon, 20 Jul 2020 19:02:55 GMT
< server: uvicorn
< content-length: 25
< content-type: application/json
< access-control-allow-credentials: true
<
* Connection #0 to host localhost left intact
{"message":"Hello World"}%
```
(as you can see, there is a `access-control-allow-credentials` header, so middleware works when `Origin` is present, but requests are never rejected)

- I am able to make any kind of request despite the fact that only `GET` is allowed
```sh
~ curl -X POST localhost:8081 -v
*   Trying 127.0.0.1:8081...
* Connected to localhost (127.0.0.1) port 8081 (#0)
> POST / HTTP/1.1
> Host: localhost:8081
> User-Agent: curl/7.71.1
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< date: Mon, 20 Jul 2020 19:01:00 GMT
< server: uvicorn
< content-length: 25
< content-type: application/json
<
* Connection #0 to host localhost left intact
{"message":"Hello World"}%
```
- `OPTIONS` requests are rejected with `405`
```sh
~ curl -X OPTIONS localhost:8081 -H 'Origin: http://localhost' -v
*   Trying 127.0.0.1:8081...
* Connected to localhost (127.0.0.1) port 8081 (#0)
> OPTIONS / HTTP/1.1
> Host: localhost:8081
> User-Agent: curl/7.71.1
> Accept: */*
> Origin: http://localhost
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 405 Method Not Allowed
< date: Mon, 20 Jul 2020 19:11:50 GMT
< server: uvicorn
< content-length: 31
< content-type: application/json
< access-control-allow-credentials: true
< access-control-allow-origin: http://localhost
< vary: Origin
<
* Connection #0 to host localhost left intact
{"detail":"Method Not Allowed"}%
```

### Environment

* OS: `Linux 5.7.7-arch1-1`
* FastAPI Version: `0.59.0`
* Python version: `3.8.3`

**Comment by Fogapod** [comment]
Looks like I completely misunderstood the CORS spec.

**Comment by tiangolo** [comment]
Thanks for reporting back and closing the issue @Fogapod  :+1: :cake:
